numpy/fft/_pocketfft_umath.c

```diff
@@ -0,0 +1,354 @@
+/*
+ * This file is part of pocketfft.
+ * Licensed under a 3-clause BSD style license - see LICENSE.md
+ */
+
+/*
+ *  Main implementation file.
+ *
+ *  Copyright (C) 2004-2018 Max-Planck-Society
+ *  \author Martin Reinecke
+ */
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#define PY_SSIZE_T_CLEAN
+#include <assert.h>
+#include <Python.h>
+
+#include "numpy/arrayobject.h"
+#include "numpy/ufuncobject.h"
+
+#include "npy_config.h"
+
+#include "pocketfft/pocketfft.h"
+
+/*
+ * Copy all nin elements of input to the first nin of the output,
+ * and any set any remaining nout-nin output elements to 0
+ * (if nout < nin, copy only nout).
+ */
+
+static inline void
+copy_data(char* in, npy_intp step_in, npy_intp nin,
+          char* out, npy_intp step_out, npy_intp nout, npy_intp elsize)
+{
+    npy_intp ncopy = nin <= nout? nin : nout;
+    if (step_in == elsize && step_out == elsize) {
+        memcpy(out, in, ncopy*elsize);
+    } else {
+        char *ip = in, *op = out;
+        for (npy_intp i = 0; i < ncopy; i++, ip += step_in, op += step_out) {
+            memcpy(op, ip, elsize);
+        }
+    }
+    if (nin < nout) {
+        char *op = out + nin*elsize;
+        if (step_out == elsize) {
+            memset(op, 0, (nout-nin)*elsize);
+        }
+        else {
+            for (npy_intp i = nin; i < nout; i++, op += step_out) {
+                memset(op, 0, elsize);
+            }
+        }
+    }
+}
+
+
+/*
+ * Loops calling the pocketfft code.
+ *
+ * Unfortunately, the gufunc machinery does not (yet?) allow forcing contiguous
+ * inner loop data, so we create a contiguous output buffer if needed
+ * (input gets copied to output before processing, so can be non-contiguous).
+ */
+static void
+fft_loop(char **args, npy_intp const *dimensions, npy_intp const *steps, void *func)
+{
+    char *ip = args[0], *fp = args[1], *op = args[2];
+    npy_intp n_outer = dimensions[0];
+    npy_intp si = steps[0], sf = steps[1], so = steps[2];
+    npy_intp nin = dimensions[1], nout = dimensions[2];
+    npy_intp isi = steps[3], iso = steps[4];
+    int (*cfft_function)(cfft_plan, double *, double) = func;
+    npy_intp npts = nout;
+    char *buffer = NULL;
+    int no_mem = 1;
+    npy_bool inplace = 0;
+
+    cfft_plan plan = make_cfft_plan(npts);
+    if (plan == NULL) {
+        goto fail;
+    }
+    if (iso != sizeof(npy_cdouble)) {
+        buffer = malloc(npts * sizeof(npy_cdouble));
+        if (buffer == NULL) {
+            goto fail;
+        }
+    }
+    else {
+        inplace = (ip == op && nin == nout && si == so && isi == iso);
```

Comment:
The strides should be guaranteed to be true I think (probably even the `nin == nout`).  If they were not, we could be in trouble!  Maybe make it an assert or even an error if you like?

Reply:
I don't think that is true generally, e.g., one could put in `fft(in, 1., out=in.T)` which would lead to the same pointer and sizes if all axes of `in` have the same size (say a square array). But the strides would differ.

Also, things like `fft(in[:nin], 1., out=in)` are in principle possible...

Reply:
The ufunc machinery should detect and prevent the overlap.  My point is that if you would pass `x, out=x[::-1]` then the copy will not work correctly anyway.

Reply:
Ah, indeed you are right! I added quite stringent tests, and those pass also by just tesing `ip == op` for in-place.